### PR TITLE
Make Kerberos authentication optional

### PR DESCRIPTION
**Pull request by shamsimam**

### CODE REVIEW

**Comment by dposada** [comment]
@shamsimam This is marked as WIP and is failing Travis -- is it ready for review?

**Comment by shamsimam** [comment]
It is WIP as I am yet to run integration tests. It is ready for feedback on the general approach. 

Unit tests pass locally, I will work on getting them to pass on Travis.

File: waiter/src/waiter/auth/authentication.clj [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+(ns waiter.auth.authentication
+  (:require [clojure.tools.logging :as log]))
+
+(def ^:const AUTH-COOKIE-NAME "x-waiter-auth")
+
+(defprotocol Authenticator
+  (auth-type [this]
+    "Returns a keyword identifying the type of authenticator.")
+  (check-user [this user service-id]
+    "Checks if the user has valid tickets under the authentication scheme. Throws an exception if not.")
```
Comment (dposada) [inline waiter/src/waiter/auth/authentication.clj]: Isn't "tickets" a Kerberos-specific term?

File: waiter/src/waiter/auth/authentication.clj [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+;; An anonymous request does not contain any authentication information.
+;; This is equivalent to granting everyone access to the resource.
+;; The anonymous authenticator attaches the prinicpal of the user running Waiter to the request.
+;; In particular, this enables requests to launch processes as the user running Waiter.
+;; Use of this authentication mechanism is strongly discouraged for production use.
+(defrecord AnonymousAuthenticator []
+  Authenticator
+  (auth-type [_]
+    :anonymous)
+  (check-user [_ _ _]
```
Comment (dposada) [inline waiter/src/waiter/auth/authentication.clj]: Nitpick: I would favor a blank line between functions

File: waiter/src/waiter/auth/authentication.clj [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    (comment "do nothing"))
+  (create-auth-handler [_ request-handler]
+    (let [process-username (System/getProperty "user.name")]
```
Comment (dposada) [inline waiter/src/waiter/auth/authentication.clj]: Instead of getting the `user.name` property, can we get the user name from the configuration?

File: waiter/src/waiter/auth/authentication.clj [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      (log/warn "use of AnonymousAuthenticator is strongly discouraged for production use:"
+                "requests will use principal" process-username)
+      (fn anonymous-handler [request]
+        (let [request' (assoc request
+                         :authorization/user process-username
+                         :authenticated-principal process-username)]
+          (request-handler request'))))))
+
+(defn anonymous-authenticator
```
Comment (dposada) [inline waiter/src/waiter/auth/authentication.clj]: Why is this needed?

File: waiter/src/waiter/auth/kerberos.clj [hunk 5]
```diff
@@ -80,3 +83,39 @@
...
+        (log/info (:message response-map) (dissoc response-map :message))
+        (throw (ex-info "No prestashed tickets available" {:message (json/write-str response-map)
+                                                           :status 403
+                                                           :suppress-logging true}))))))
+
+(defrecord KerberosAuthenticator [password query-chan]
+  auth/Authenticator
+  (auth-type [_]
+    :kerberos)
+  (check-user [_ user service-id]
```
Comment (dposada) [inline waiter/src/waiter/auth/kerberos.clj]: Nitpick: I would favor a blank line between functions

File: waiter/src/waiter/client_tools.clj [hunk 6]
```diff
@@ -34,6 +34,8 @@
 (def ^:const WAITER-PORT 9091)
 (def ^:const HTTP-SCHEME "http://")
 
+(def ^:const use-spnego (-> (System/getenv "USE_SPNEGO") str Boolean/parseBoolean))
```
Comment (dposada) [inline waiter/src/waiter/client_tools.clj]: What happens if this environment variable is not set at all? We should support not set being the same as false

File: waiter/src/waiter/core.clj [hunk 7]
```diff
@@ -694,7 +676,7 @@
                          (let [local-router (InetAddress/getLocalHost)
                                waiter-router-hostname (.getCanonicalHostName local-router)
                                waiter-router-ip (.getHostAddress local-router)]
-                           (waiter-request?-factory #{hostname waiter-router-hostname waiter-router-ip})))})
+                           (waiter-request?-factory (set [hostname waiter-router-hostname waiter-router-ip]))))})
```
Comment (dposada) [inline waiter/src/waiter/core.clj]: Why the change here?

File: waiter/src/waiter/client_tools.clj (hunk 6 above)
Comment (shamsimam) [inline waiter/src/waiter/client_tools.clj]: Not set is being interpreted as false as a blank string is parsed as false.

File: waiter/src/waiter/core.clj (hunk 7 above)
Comment (shamsimam) [inline waiter/src/waiter/core.clj]: `waiter-router-hostname` and `waiter-router-ip` are the same value when I run locally and this causes an error with the set literal.

File: waiter/src/waiter/auth/kerberos.clj (hunk 5 above)
Comment (shamsimam) [inline waiter/src/waiter/auth/kerberos.clj]: Done.

File: waiter/src/waiter/auth/authentication.clj (hunk 4 above)
Comment (shamsimam) [inline waiter/src/waiter/auth/authentication.clj]: Factory function for the anonymous authenticator.

File: waiter/src/waiter/auth/authentication.clj (hunk 2 above)
Comment (shamsimam) [inline waiter/src/waiter/auth/authentication.clj]: Done.

File: waiter/src/waiter/auth/authentication.clj (hunk 1 above)
Comment (shamsimam) [inline waiter/src/waiter/auth/authentication.clj]: Changed the doctring.

File: waiter/config-full.edn [hunk 8]
```diff
@@ -144,10 +144,26 @@
...
+ ;; Waiter supports authenticating requests to identify the principal of the requesting user.
+ ;; The principal is then used to validate access to services and also to launch services.
+ :authenticator-config {
+                        ;; :kind :anonymous allows anonymous access to services (for testing purposes only):
+                        :kind :anonymous
+                        :anonymous {;; Custom implementations should specify a :factory-fn
+                                    ;; that returns an instance of waiter.auth.authentication.Authenticator:
+                                    :factory-fn waiter.auth.authentication/anonymous-authenticator}
+
+                        ;; :kind :anonymous allows anonymous access to services:
```
Comment (shamsimam) [inline waiter/config-full.edn]: Done.

File: waiter/src/waiter/auth/authentication.clj (hunk 3 above)
Comment (shamsimam) [inline waiter/src/waiter/auth/authentication.clj]: I introduced a `launch-as-user` config for the anonymous authenticator. By default, it is configured to the `user.name` property.

**Comment by shamsimam** [comment]
@dposada incorporated your feedback

File: waiter/src/waiter/settings.clj [hunk 9]
```diff
@@ -152,7 +153,10 @@
   (utils/map->json-response (into (sorted-map) (sanitize-settings settings))))
 
 (def settings-defaults
-  {:cors-config {:kind :patterns
+  {:authenticator-config {:kind :anonymous
+                          :anonymous {:factory-fn 'waiter.auth.authentication/anonymous-authenticator
+                                      :launch-as-user (System/getProperty "user.name")}}
```
Comment (dposada) [inline waiter/src/waiter/settings.clj]: I would favor making `:launch-as-user` a required field when using `:anonymous` as opposed to defaulting it to the `user.name` property

Comment (shamsimam) [inline waiter/src/waiter/settings.clj]: That would then mean the default settings is insufficient to launch waiter. Are we okay to make a config file required?

Comment (dposada) [inline waiter/src/waiter/settings.clj]: Isn't a config file already required? That's why we have `config-minimal.edn`

Comment (shamsimam) [inline waiter/src/waiter/settings.clj]: Comment retracted :) Will make launch-user a required config
